### PR TITLE
Feat/codecarbon

### PR DESCRIPTION
**Pull request by AJDERS**
This PR introduces:

- `codecarbon` usage.
- metric configs corresponding to `carbon_emissions` and `energy_consumed` (`EMISSIONS` and `POWER` resp.) are added to `task_configs.py`. **We should consider creating `metric_configs.py` no?**
- function for checking if internet connection is available.

Ill briefly describe how the carbon emission tracking works:

1. During the preparation of the tracker, internet availability is checked. The tracker is instantiated in either online or offline mode.
2. The tracker is started just before the loop over the dataloader and is stopped immediately after.
3. The `carbon_emissions` and `energy_consumed` for the iteration are stored.
4. Before logging a copy of `self.task_config.metrics` is created and the metrics `EMISSIONS` and `POWER` are appended to ensure:
    * The logging is done in the same way across all metrics, independent of when they are calculated,
    * The `self.task_config.metrics` is not mutated, so in case of multiple `model_ids` the metrics `EMISSIONS` and `POWER` will not be used in `_compute_metrics` and cause errors.

It might be that the copying of `self.task_config.metrics` is unnecessary, i need to investigate further.

### CODE REVIEW

File: src/aiai_eval/cli.py [hunk 1]
```diff
@@ -76,6 +76,20 @@
     show_default=True,
     help="Whether extra input should be outputted during benchmarking",
 )
+@click.option(
+    "--track-carbon-emissions",
+    "-tce",
+    is_flag=True,
+    show_default=True,
+    help="Whether to track carbon usage.",
```
Comment (saattrupdan) [inline src/aiai_eval/cli.py:84]: Maybe add here that `country-iso-code` should be specified?

File: src/aiai_eval/cli.py [hunk 2]
```diff
@@ -76,6 +76,20 @@
...
+)
+@click.option(
+    "--country-iso-code",
+    "-co",
+    default="DNK",
```
Comment (saattrupdan) [inline src/aiai_eval/cli.py:89]: I think this is potentially misleading, if they forget about this argument. Maybe set default to be an empty string, and raise an error if `--track-carbon-emissions` is set and this is not?

File: src/aiai_eval/cli.py [hunk 3]
```diff
@@ -76,6 +76,20 @@
...
+    show_default=True,
+    help="The 3-letter alphabet ISO Code of the country where the compute infrastructure is hosted.",
```
Comment (saattrupdan) [inline src/aiai_eval/cli.py:91]: Adding a link to a list of all such country codes would be helpful.

File: src/aiai_eval/config.py [hunk 4]
```diff
@@ -144,6 +150,8 @@ class EvaluationConfig:
     save_results: bool
     verbose: bool
     testing: bool = False
+    track_carbon_emissions: bool = False
+    country_iso_code: str = "DNK"
```
Comment (saattrupdan) [inline src/aiai_eval/config.py:154]: Same thing as above regarding default value

File: src/aiai_eval/config.py [hunk 5]
```diff
@@ -135,6 +135,12 @@ class EvaluationConfig:
             Whether to print verbose output.
         testing (bool, optional):
             Whether a unit test is being run. Defaults to False.
+        track_carbon_usage (bool):
+            Whether to track carbon usage.
+        country_iso_code (str):
+            The 3-letter alphabet ISO Code of the country where the compute infrastructure
+            is hosted. This is used when tracking carbon usage.
```
Comment (saattrupdan) [inline src/aiai_eval/config.py:142]: Same thing as above regarding link to list of country codes.

File: src/aiai_eval/task.py [hunk 6]
```diff
@@ -641,6 +664,55 @@ def _adjust_label_ids(
...
+        tracker_name = self.task_config.dataset_name
+        if self.evaluation_config.verbose:
+            log_level = "info"
+        else:
+            log_level = "error"
+
+        if internet_connection_available():
+            carbon_tracker = EmissionsTracker(
+                project_name=tracker_name,
+                measure_power_secs=5,
```
Comment (saattrupdan) [inline src/aiai_eval/task.py:686]: What does this hyperparameter do, and should a user be able to tweak it?

File: src/aiai_eval/task.py [hunk 7]
```diff
@@ -641,6 +664,55 @@ def _adjust_label_ids(
...
+            country_iso_code = self.evaluation_config.country_iso_code
+            if country_iso_code == "DNK":
+                message = (
+                    "The carbon tracker calculates carbon usage based on power consumption, "
+                    "and the country where the compute infrastructure is hosted. Internet connection "
+                    "was not available and hence the location of the infrastructure could not be "
+                    "automatically fetched, because of the this the location defaults to 'DNK'."
+                    "If the compute infrastructure is not hosted in Denmark, change the"
+                    "'country_iso_code' to the correct ISO code."
+                )
```
Comment (saattrupdan) [inline src/aiai_eval/task.py:703]: If we change the default country code then this check should be changed accordingly.

File: src/aiai_eval/utils.py [hunk 8]
```diff
@@ -114,8 +116,6 @@ def log_scores(
     for metric_cfg in metric_configs:
         agg_scores = aggregate_scores(scores=scores, metric_config=metric_cfg)
         test_score, test_se = agg_scores["test"]
-        test_score *= 100
-        test_se *= 100
```
Comment (saattrupdan) [inline src/aiai_eval/utils.py:118]: Don't like scores of the form XX.XX%?

File: src/aiai_eval/utils.py [hunk 9]
```diff
@@ -114,8 +116,6 @@ def log_scores(
...
 
         msg = f"{metric_cfg.pretty_name}:\n  - Test: {test_score:.2f} ± {test_se:.2f}"
```
Comment (saattrupdan) [inline src/aiai_eval/utils.py:120]: If we don't multiply the scores by 100 then this should be changed to .4f, as two decimals won't give enough detail; it'll only state things like 0.54 instead of 0.5423.

File: src/aiai_eval/utils.py [hunk 10]
```diff
@@ -164,3 +164,12 @@ def aggregate_scores(
         results["test"] = (test_score, 1.96 * test_se)
 
         return results
+
+
+def internet_connection_available() -> bool:
+    try:
+        # Check if internet connection is available.
```
Comment (saattrupdan) [inline src/aiai_eval/utils.py:171]: Redundant comment; can be moved to docstring.

**Review by saattrupdan — CHANGES_REQUESTED** [review]

**Comment by saattrupdan** [comment]
Also, a `metric_configs.py` might be useful yeah. But let's leave that for another PR.

File: src/aiai_eval/utils.py (hunk 9 above)
Comment (AJDERS) [reply src/aiai_eval/utils.py:120]: Agreed.

File: src/aiai_eval/utils.py (hunk 8 above)
Comment (AJDERS) [reply src/aiai_eval/utils.py:118]: It does not make sense to multiply the emissions/power metrics by 100. Thats why i removed them.

File: src/aiai_eval/cli.py (hunk 1 above)
Comment (AJDERS) [reply src/aiai_eval/cli.py:84]: Seems reasonable. Will do.

File: src/aiai_eval/cli.py (hunk 2 above)
Comment (AJDERS) [reply src/aiai_eval/cli.py:89]: Will do.

File: src/aiai_eval/task.py (hunk 6 above)
Comment (AJDERS) [reply src/aiai_eval/task.py:686]: Its how often the power usage is checked in seconds. I think it should not be tweakable as it could be missused to obtain an artificially low emission. So i kept it hardcoded for better comparison between models.

File: src/aiai_eval/utils.py (hunk 8 above)
Comment (saattrupdan) [reply src/aiai_eval/utils.py:118]: Ah, fair enough!

**Review by saattrupdan — APPROVED** [review]
LGTM!
